### ISSUE TITLE
Support modulo by Euclidean divisor

### ISSUE THREAD

**Issue by Paalon** [nim-lang/Nim#7910]
Nim's `mod` is modulo by truncated divisor but Euclidean divisor has many good mathematical properties than truncated divisor and more convenient in many applications.
https://en.wikipedia.org/wiki/Modulo_operation
I think Nim should have modulo based on Euclidean divisor, but there are some problems.
- There are quite few languages which have both modulo and adopt truncated divisor modulo as `mod` (according to wikipedia's page, only Fortran).
- Especially, most of the scientific computing programming languages have Euclidean divisors as `mod` (MATLAB, Maxima, R, Julia) or `Mod` (Mathematica) or  `modulo` (Fortran).
- There are many languages they adopt Euclidean divisors as `mod`, truncated divisors as `rem` (Ada, Closure, Common Lisp, Haskell, Prolog)

I like `mod` as Euclidean divisor, but I think discussions are required.

**Comment by GULPF** [comment]
There are actually three different ways to implement `div`/`mod`: floor, trunc and euclidean ([here's a good reference](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/divmodnote-letter.pdf)). In fact, I think that several of the languages that you say use euclidean modulo actually use floor modulo.

In my opinion, the behavior of the `mod` and `div` operators should not be changed since the trunc based variant is common amongst system programming languages, and I believe it's a lot faster on a typical CPU. Perhaps more importantly, I believe Nim should avoid making breaking changes in core keywords, otherwise it will never reach the mythical version 1.

Instead, the other variants should be available as procs in the `math` module. I've already added procs for floor based `div`/`mod` in #7118, so we're only missing the euclidean implementations.

**Comment by Paalon** [comment]
Sorry for confusing between Euclidean and floor. Here is a list of languages what kind of symbol is used for modulo.

| Released | Language      | `%`         | `mod`       | `rem`       | `modulo` | `remainder` | `%%`              | `Mod`                   | `fmod`                  | `floorMod`                        |
| -------- | ------------- | --------- | --------- | --------- | ------ | --------- | --------------- | --------------------- | --------------------- | ------------------------------- |
| 1957     | Fortran       | -         | truncated | -         | floor  | -         | -               | -                     | -                     | -                               |
| 1970     | Pascal        | -         | truncated | -         | -      | -         | -               | -                     | -                     | -                               |
| 1972     | C             | truncated | -         | -         | -      | -         | -               | -                     | -                     | -                               |
| 1975     | Scheme (R7RS) | -         | -         | -         | floor  | truncated | -               | -                     | -                     | -                               |
| 1982     | Maxima        | -         | floor     | -         | -      | -         | -               | -                     | -                     | -                               |
| 1983     | Ada           | -         | floor     | truncated | -      | -         | -               | -                     | -                     | -                               |
| 1983     | C++           | truncated | -         | -         | -      | -         | -               | -                     | -                     | -                               |
| 1984     | MATLAB        | -         | floor     | truncated | -      | -         | -               | -                     | -                     | -                               |
| 1984     | Common Lisp   | -         | floor     | truncated | -      | -         | -               | -                     | -                     | -                               |
| 1985     | Scheme (Chez) | -         | Euclidean | -         | floor  | truncated | -               | -                     | -                     | -                               |
| 1990     | Haskell       | -         | floor     | truncated | -      | -         | -               | -                     | -                     | -                               |
| 1991     | Python        | floor     | -         | -         | -      | -         | -               | -                     | truncated (for float) | -                               |
| 1995     | Java          | truncated | -         | -         | -      | -         | -               | -                     | -                     | floor                           |
| 1995     | JavaScript    | truncated | -         | -         | -      | -         | -               | -                     | -                     | -                               |
| 1995     | Ruby          | floor     | -         | -         | floor  | truncated | -               | -                     | -                     | -                               |
| 1996     | R             | -         | -         | -         | -      | -         | floor           | -                     | -                     | -                               |
| 2000     | C#            | truncated | -         | -         | -      | -         | -               |                       |                       |                                 |
| 2003     | Scala         | truncated | -         | -         | -      | -         | -               | -                     | -                     | -                               |
| 2006     | Rust          | truncated | -         | -         | -      | -         | -               | -                     | -                     | -                               |
| 2007     | Clojure       | -         | floor     | truncated | -      | -         | -               | -                     | -                     | -                               |
| 2008     | Nim (0.18)    | -         | truncated | -         | -      | -         | (only for uint) | -                     | -                     | (recently implemented as floor) |
| 2008     | Go            | truncated | -         | -         | -      | -         | -               | truncated (for float) | -                     | -                               |
| 2012     | Julia         | truncated | floor     | truncated | -      | -         | -               | -                     | -                     | -                               |
| 2012     | Crystal       | floor     | -         | -         | -      | -         | -               | -                     | -                     | -                               |
| 2014     | Swift         | truncated | -         | -         | -      | -         | -               | -                     | -                     | -                               |

**Comment by Paalon** [comment]
I agree that many system programming languages have truncated mod but the name `mod` for truncated is not common. I don't think the name `floorMod` is friendly for users. It's one of the essential function but too long name.

**Comment by mratsim** [comment]
`%` is read `mod` in C, C++, Java and a couple other languages so it is actually common.

Quoting Wikipedia:
> In computing, the modulo operation finds the **remainder** after division of one number by another (sometimes called modulus).

Emphasis mine, `programming modulo` == `mathematical remainder`.

**Comment by andreaferretti** [comment]
Just to be sure: the only ambiguity here is the representative for the modulo class in the negative case, or is there something more?

**Comment by mratsim** [comment]
yes only the negative case. Otherwise all 3 (truncated, floored, euclidean) give the same result.

**Comment by dom96** [comment]
It's far too late to change this now. I would argue that everything that came after Pascal is wrong, since Pascal should set the trend and Nim is influenced by it which explains why ``mod`` is used instead of ``%`` (like in Python).

We can add Euclidean divisor using a different name or operator.

**Comment by Araq** [comment]
Most of these pascal-inspired operators are also keywords so that operators like ``%`` and ``<<`` are available for more common operations. I've argued this before, many programs use JSON and no math related stuff beyond simple additions.
